### ISSUE TITLE
Potential usability problems that follow from users likely expecting type inference where it doesn't exist

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#31941]
#30177 revealed that this is a (somewhat?) common mistake:

```
myTopLevel() => someCalculation();
```

This *looks* like it has type inference. But its a top level, so it doesn't. Specifically in regards to to #30177, this is an error if `someCalculation` returns `void`, which is becoming an error which we have to find & correct.

I bet the accidental type mismatches here go far beyond `void`/`dynamic` mismatches. There are likely places where people are returning complex types, and they don't realize they have thrown type safety out of the window.

I made this issue here, rather than there, to discuss if this really is a general usability problem, and if there's a general solution.

I have three thoughts which I'm sure are each terrible:

1. Obviously there are many cases where type inference cannot work easily at the top level, such as `myTopLevel() => myTopLevel()`. But is it worth exploring making arrow functions try inference when the return type is omitted? And maybe err, rather than using `dynamic` when inference fails?

But that's a huge can of worms. If the underlying problem is that people expect inference where it doesn't exist, it may exacerbate the problem to support it here.

2. Larger outlook still, maybe dynamic shouldn't be allowed to be used implicitly here at all. How hard is it to write `void` if its void, `dynamic` if its dynamic, or a `type`, which is the most work but also the case where accidentally getting `dynamic` is the largest problem.

3. or, (getting really really weird here) maybe implicit casting to dynamic should only occur when the dynamic was explicit. As in, the use cases for dynamic typing and strict typing are very different and it should maybe be never possible to go from one world to the other without some explicit casts/typing. Defining "some" is obviously a daunting task.

In this case, the user never used the word `dynamic` so maybe we should assume that they intended to have type safety. In that case, I imagine this to be a warning where a non-dynamic type was *implicitly* upcast into an *implicit* dynamic return type.. However, `dynamic foo() => bar()` would be able to implicitly upcast to *explicit* dynamic, and `foo() => bar() as dynamic` or `foo() => dynamicFn()` wouldn't require an implicit cast at all, so it would pass OK.

I imagine none of these three ideas were new, and all of them have been discarded for good reason. I don't expect a detailed reply on why my three ideas aren't practical/useful/good.

However, are there any other ideas that are useful? Or are these ideas that have not been ruled out yet that warrant further discussion?

**Comment by eernstg** [comment]
I think we shouldn't infer type annotations of formal parameters of anything else than function literals. That is, no inference for the type of `p` in `foo(p) => e` nor `bar(p) { S }`.

I believe that we maintain this rule consistently, but @leafpetersen can confirm.

However, as a special deviation from this "don't infer function signatures" meta-rule, the _return type of a local function_ is inferred if omitted.

One main difference between local and global functions is that the global ones may be mutually dependent on each other, even across library boundaries. This means that inference of such a return type is in a completely different complexity class than inference of the return type of a local function. This could motivate the difference, even though I do recognize that it is a source of confusion.

So we have a migration cost to worry about; but apart from that, wouldn't it work OK to stick to the current rules? (such that `myTopLevel() => someCalculation();` is an error when `someCalculation` has return type `void`, because there is no exception allowing void evaluation at that position).

I think it may be helpful for developers to _read_ `void myTopLevel() => someCalculation();` rather than the version with no return type, and it would in any case be a "good style" habit to specify the return type for all top level functions in strictly typed code.

If so, we're talking about a migration cost which is to some extent justified by the outcome, not just a silly "whatever, we must write that kind of function like that now" migration cost. ;)

**Comment by MichaelRFairhurst** [comment]
> So we have a migration cost to worry about; but apart from that, wouldn't it work OK to stick to the current rules? (such that myTopLevel() => someCalculation(); is an error when someCalculation has return type void, because there is no exception allowing void evaluation at that position).

Yes absolutely. In this case, though, I opened the ticket to handle the cases that aren't being caught, such as:

```
commonlyUsedFunction() => strictlyTypedFunction();
```

I could imagine, for instance, a case where the toplevel `commonlyUsedFunction()` originally had a well defined type, and then someone saw it and incorrectly thought to themselves "this could be inferred!" They might then successfully land a CL that removes the type annotation, because it would introduce no compile-time errors to make such a switch.

Next thing you know, people wonder why

```
  var x = commonlyUsedFunction();
  x.invalidMembersAndUsageAndStuff;
```

no longer performs any checks for them.

#30177 won't, unfortunately, catch this case at all.

**Comment by eernstg** [comment]
OK, I misunderstood, I thought it was mainly about `void`. I do agree that it is confusing and error-prone to have these functions that look syntactically identical as local functions and as top-level functions, and the former will have their return type inferred and the latter will have `dynamic`.

It might help developers who are already using a rather strict style of typing that the top-level functions can be flagged by analysis with `--no-implicit-dynamic`.

It may not be so easy to extend inference, because it is potentially quite expensive to perform the top-level inference if it is made significantly more powerful, and inference of signatures for mutually dependent declarations may push the problem into a complexity class which is intractable (but @leafpetersen knows the inference algorithm better than I do, so I'll just say these things vaguely ;-).
